addons/udes_security/models/models.py

```diff
@@ -0,0 +1,69 @@
+from lxml import etree
+
+from odoo import models, api
+from odoo.osv.orm import setup_modifiers
+
+
+class BaseModel(models.AbstractModel):
+    _inherit = "base"
+
+    @api.model
+    def fields_view_get(self, view_id=None, view_type="form", toolbar=False, submenu=False):
+        """Override to remove buttons/editable functionality for view only desktop users"""
+
+        def _remove_buttons(doc):
+            """Remove all buttons from the view"""
+            for node in doc.xpath("//button"):
+                node.getparent().remove(node)
+
+        def _set_fields_readonly(doc, res):
+            """Set all fields in the view to readonly"""
+            for node in doc.xpath("//field"):
+                field_name = node.get("name")
+                if field_name:
+                    node.set("readonly", "1")
+                    setup_modifiers(node, res["fields"][field_name])
+
+        res = super().fields_view_get(
+            view_id=view_id, view_type=view_type, toolbar=toolbar, submenu=submenu
+        )
+
+        if self.env.user.u_desktop_readonly:
+            doc = etree.XML(res["arch"])
+
+            doc.set("create", "false")
+            doc.set("edit", "false")
+            doc.set("duplicate", "false")
+            doc.set("delete", "false")
+
+            if doc.tag == "form":
+                _remove_buttons(doc)
+                # Make all fields readonly
+                # Readonly user shouldn't be able to get a form into edit mode, but it is possible
+                # via the URL to bring up new record screen and quick create for relational fields
+                _set_fields_readonly(doc, res)
+
+            elif doc.tag == "tree":
+                _remove_buttons(doc)
+                _set_fields_readonly(doc, res)
+                # Sequence handle allows records to be re-ordered
+                for node in doc.xpath("//field[@widget='handle']"):
```

Comment:
We could generalise `_remove_buttons(doc)` into `_remove_nodes(doc, xpath)` and reuse it for this loop and the next one.

Reply:
Yep good idea :) 

I kept `_remove_buttons` (needed also to remove from Kanban views, as buttons can lead to changes, e.g. Install on the module cards...) but this just makes use of `_remove_nodes`. :)